Update EIP-4788: initial stab at v2 (#7456)
ethereum/EIPs: EIPS/eip-4788.md

## Specification

| constants                    | value                                        |
|---                           |---                                           |
| `FORK_TIMESTAMP`             | TBD                                          |
| `HISTORICAL_ROOTS_MODULUS`   | `98304`                                      |
| `SYSTEM_ADDRESS`             | `0xfffffffffffffffffffffffffffffffffffffffe` |

> | `BEACON_ROOTS_ADDRESS`       | `0xBeAC00541d49391ED88ABF392bfC1F4dEa8c4143` |

Reviewer: ```suggestion
| `BEACON_ROOTS_ADDRESS`       | `0xbEac00dDB15f3B6d645C48263dC93862413A222D` |
```

##### `set`

* Caller provides the parent beacon block root as calldata to the contract.
* Set the storage value at `header.timestamp % HISTORICAL_ROOTS_MODULUS` to be `header.timestamp`
* Set the storage value at `header.timestamp % HISTORICAL_ROOTS_MODULUS + HISTORICAL_ROOTS_MODULUS` to be `calldata[0:32]`

##### Pseudocode

```python
if evm.caller == SYSTEM_ADDRESS:
    set()
else:
    get()

def get():
    if len(evm.calldata) != 32:
        evm.revert()

    timestamp_idx = to_uint256_be(timestamp) % HISTORICAL_ROOTS_MODULUS
    timestamp = storage.get(timestamp_idx)

    if timestamp != evm.calldata:
        evm.revert()

    root_idx = timestamp_idx + HISTORICAL_ROOTS_MODULUS
    root = storage.get(root_idx)
    
    evm.return(root)

def set():
    timestamp_idx = to_uint256_be(evm.timestamp) % HISTORICAL_ROOTS_MODULUS
    root_idx = timestamp_idx + HISTORICAL_ROOTS_MODULUS

    storage.set(timestamp_idx, evm.timestamp)
    storage.set(root_idx, evm.calldata)
```

##### Bytecode

The exact initcode to deploy is shared below.

> ```asm

Reviewer 2: i think it would be really nice and convenient to have bytecode hex as well for below to leave no room for any mismatch in client

For a genesis block with no existing parent beacon block root the 32 zero bytes are used as a root placeholder.

#### Beacon roots contract

The beacon roots contract has two operations: `get` and `set`. The input itself is not used to determine which function to execute, for that the result of `caller` is used. If `caller` is equal to `SYSTEM_ADDRESS` then the operation to perform is `set`. Otherwise, `get`.

##### `get`

* Callers provide the `timestamp` they are querying encoded as 32 bytes in big-endian format.
* If the input is not exactly 32 bytes, the contract must revert.
* Given `timestamp`, the contract computes the storage index in which the timestamp is stored by computing the modulo `timestamp % HISTORICAL_ROOTS_MODULUS` and reads the value.
* If the `timestamp` does not match, the contract must revert.

> * Finally, the beacon root associated with the timestamp is accessed at returned to the user. It is stored at `timestamp % HISTORICAL_ROOTS_MODULUS + HISTORICAL_ROOTS_MODULUS`.

Reviewer 3: ?

```suggestion
* Finally, the beacon root associated with the timestamp is returned to the user. It is stored at `timestamp % HISTORICAL_ROOTS_MODULUS + HISTORICAL_ROOTS_MODULUS`.
```

    
    evm.return(root)

def set():
    timestamp_idx = to_uint256_be(evm.timestamp) % HISTORICAL_ROOTS_MODULUS
    root_idx = timestamp_idx + HISTORICAL_ROOTS_MODULUS

    storage.set(timestamp_idx, evm.timestamp)
    storage.set(root_idx, evm.calldata)
```

##### Bytecode

The exact initcode to deploy is shared below.

```asm
push1 0x5a
dup1
push1 0x09
push0
codecopy
push0
return

caller
push20 0xfffffffffffffffffffffffffffffffffffffffe
eq
push1 0x42
jumpi

push1 0x20
calldatasize
eq
push1 0x24
jumpi

push0
push0
revert

jumpdest
push3 0x018000
push0
calldataload
mod
dup1
sload
push0
calldataload
eq
iszero
push1 0x3d
jumpi

push3 0x018000
add
sload
push0
mstore

jumpdest
push1 0x20
push0
return

jumpdest
timestamp
push3 0x018000
timestamp
mod
sstore
push0
calldataload
push3 0x018000
timestamp
mod
push3 0x018000
add
sstore
stop
```

#### Deployment

The beacon roots contract is deployed like any other smart contract. A special synthetic address is generated
by working backwards from the desired deployment transaction:

```json
{
  "type": "0x2",
  "chainId": "0x1",
  "nonce": "0x0",
  "to": null,
  "gas": "0xd4f8",
  "gasPrice": null,
  "maxPriorityFeePerGas": "0x9c7652400",

>   "maxFeePerGas": "0xe8d4a51000",

Reviewer 3: where do these gas values come from?

to echo what others have said, it seems fragile to enshrine a "synthetic transaction"

easier to just drop byte code a la EIP-1011

Author: Just arbitrary values that would be included on mainnet. It isn't really enshrining the tx. I would think more of it as two steps: 1) deploy contract 2) set address in clients to call as part of pre block processing. A synthetic tx is just one way of deploying the contract. I think it is best because it is clear, transparent, and permissionless. We could just as easily have someone hand deploy the contract before the fork (now even, if the byte code is ready) and then set the address in client configurations.

  }
```

The sender of the transaction can be calculated as `0x01d0610058aC7AEF1887d8877ee7f04B7645Dc95`. The address of the first contract deployed from the account is `rlp([sender, 0])` which equals `0x502E02F5d91024A9AF0aB81fbF0a47Eb99a013aE`. Although this style of contract creation is not tied to any specific initcode like create2 is, the synthetic address is cryptographically bound to the input data of the transaction (e.g. the initcode).

### Block processing

At the start of processing any execution block where `block.timestamp >= FORK_TIMESTAMP` (i.e. before processing any transactions), call `BEACON_ROOTS_ADDRESS` as `SYSTEM_ADDRESS` with the 32-byte input of `header.parent_beacon_block_root`. This will trigger the `set()` routine of the beacon roots contract. This is a system operation and therefore:

* the call must execute to completion, therefore the available gas can be considered as infinite
* the call does not count against the block's gas limit
* the call does not follow the [EIP-1559](./eip-1559.md) burn semantics - no value should be transferred as part of the call
* if no code exists at `BEACON_ROOTS_ADDRESS`, the call must fail silently

> Client may decide to omit an explicit EVM call and directly set the storage values.

Reviewer 4: This would make any potential modifications to the contract in the future harder or support some other contract on a different chain.

Reviewer: Yes. This sentence should have some YMMV warnings: doing this may be a workable implementation of the mainnet contract, but might lead to issues on other networks, if some other contract is used. 

(The case where no contract at all exists should be fine, since setting the values directly doesn't affect the state -- the slots are cleaned up due to the `empty`-rule at the end of block anyway. )

Perhaps something like this? 

```suggestion
Clients may decide to omit an explicit EVM call and directly set the storage values. Note: While this is a valid optimization for Ethereum mainnet, it could be problematic on non-mainnet situations in case a different contract is used.
```

> Client may decide to omit an explicit EVM call and directly set the storage values.

Reviewer 5: I think that we have to decide whether we want to do the EVM call or to directly set the storage values. I understand that when we set the value directly and the the account is empty we do remove the account at the end of the block, but there are the edge case where someone puts some eth into the account, which means it is not empty anymore. 
I would prefer to set the values directly, as that is 10x faster and we do not require all clients to implement the system operation.

Reviewer: > I think that we have to decide whether we want to do the EVM call

The consensus-definition is that an EVM call is made. A client implementor may choose to optimize this, and do a direct update instead. This choice may cause incompatibilities in non-mainnet scenarios, e.g. if the code is replaced with different code. 

> I understand that when we set the value directly and the the account is empty we do remove the account at the end of the block, but there are the edge case where someone puts some eth into the account, which means it is not empty anymore.

No, there is now code on the account, so it would not be removed, since it is not `empty`, according to the definition of `empty`. 

> I would prefer to set the values directly, as that is 10x faster and we do not require all clients to implement the system operation.

You are at liberty to do so, since "we do not require all clients to implement the system operation" -- however, the **consensus correct behaviour, in any given scenario, is given by the system-call behaviour**

IMO this is a pretty good middle ground.

```

##### Bytecode

The exact initcode to deploy is shared below.

```asm
push1 0x58
dup1
push1 0x09
push0
codecopy
push0
return

caller
push20 0xfffffffffffffffffffffffffffffffffffffffe
eq
push1 0x44
jumpi

push1 0x20
calldatasize
eq
push1 0x24
jumpi

push0
push0
revert

jumpdest
push3 0x018000
push0
calldataload
mod
dup1
sload
push0
calldataload
eq
push1 0x37
jumpi

push0
push0
revert

jumpdest
push3 0x018000
add
sload
push0
mstore
push1 0x20
push0
return

jumpdest
push3 0x018000
timestamp
mod
timestamp
dup2
sstore
push0
calldataload
swap1
push3 0x018000
add
sstore
stop
```

#### Deployment

The beacon roots contract is deployed like any other smart contract. A special synthetic address is generated
by working backwards from the desired deployment transaction:

```json
{
  "type": "0x0",
  "nonce": "0x0",
  "to": null,
  "gas": "0x27eac",
  "gasPrice": "0xe8d4a51000",
  "maxPriorityFeePerGas": null,
  "maxFeePerGas": null,
  "value": "0x0",
  "input": "0x60588060095f395ff33373fffffffffffffffffffffffffffffffffffffffe14604457602036146024575f5ffd5b620180005f350680545f35146037575f5ffd5b6201800001545f5260205ff35b6201800042064281555f359062018000015500",
  "v": "0x1b",
  "r": "0x539",

>   "s": "0x1337",

Reviewer: Alternative suggestion: 
```suggestion
  "s": "0x133700cb4fbb1368012b31",
```

Then we get a `bEaC0` -prefixed address `0xbEaC0B09a6ed668e6BC5d29e29013aCb9286bF5F` with the transaction
```json
{
  "type": "0x0",
  "nonce": "0x0",
  "to": null,
  "gas": "0x27eac",
  "gasPrice": "0xe8d4a51000",
  "maxPriorityFeePerGas": null,
  "maxFeePerGas": null,
  "value": "0x0",
  "input": "0x60588060095f395ff33373fffffffffffffffffffffffffffffffffffffffe14604457602036146024575f5ffd5b620180005f350680545f35146037575f5ffd5b6201800001545f5260205ff35b6201800042064281555f359062018000015500",
  "v": "0x1b",
  "r": "0x539",
  "s": "0x133700cb4fbb1368012b31",
  "hash": "0xa71141ed853de6fdd7c9d9355293f255011e9e512cb833a03b3ff20d8045f5b9"
}
```

Reviewer: Update: `BeAC00` - prefix:

Address: `0xBeAC00541d49391ED88ABF392bfC1F4dEa8c4143`
Tx:
```json
{
  "type": "0x0",
  "nonce": "0x0",
  "to": null,
  "gas": "0x27eac",
  "gasPrice": "0xe8d4a51000",
  "maxPriorityFeePerGas": null,
  "maxFeePerGas": null,
  "value": "0x0",
  "input": "0x60588060095f395ff33373fffffffffffffffffffffffffffffffffffffffe14604457602036146024575f5ffd5b620180005f350680545f35146037575f5ffd5b6201800001545f5260205ff35b6201800042064281555f359062018000015500",
  "v": "0x1b",
  "r": "0x539",
  "s": "0x1337007db5d301ac899c9e",
  "hash": "0x6923750e3b16ef9bfe1a70820d0d0628d7d4e84906b19ef91840fe913d688759"
}
```

Author: amazing!

Author: Ah wait did you intend for this to be the sender of the transaction or the address of the deployed contract? Because I am recovering `0xBeAC00541d49391ED88ABF392bfC1F4dEa8c4143` as the EOA address and `0xC221266EE238DbA34c6df82f75516a27087C0Bd9` as the deployed address.

Reviewer: Ah silly me. Let me get back to you on this

Reviewer: How about this then? 
_deploy_ address: `0xbEac00dDB15f3B6d645C48263dC93862413A222D`

```json
{
  "type": "0x0",
  "nonce": "0x0",
  "to": null,
  "gas": "0x27eac",
  "gasPrice": "0xe8d4a51000",
  "maxPriorityFeePerGas": null,
  "maxFeePerGas": null,
  "value": "0x0",
  "input": "0x60588060095f395ff33373fffffffffffffffffffffffffffffffffffffffe14604457602036146024575f5ffd5b620180005f350680545f35146037575f5ffd5b6201800001545f5260205ff35b6201800042064281555f359062018000015500",
  "v": "0x1b",
  "r": "0x539",
  "s": "0x133700f3a77843802897db",
  "hash": "0x14789a20c0508b81ab7a0287a12a3a41ca960aa18244af8e98689e37ed569f07"
}
```

  "hash": "0x14789a20c0508b81ab7a0287a12a3a41ca960aa18244af8e98689e37ed569f07"
}
```

> The sender of the transaction can be calculated as `0x3A700F76Ddec97c4640FdAF5a0fF1632E50A10D8`. The address of the first contract deployed from the account is `rlp([sender, 0])` which equals `0xBeAC00541d49391ED88ABF392bfC1F4dEa8c4143`. This is how `BEACON_ROOTS_ADDRESS` is determined. Although this style of contract creation is not tied to any specific initcode like create2 is, the synthetic address is cryptographically bound to the input data of the transaction (e.g. the initcode).

Reviewer: ```suggestion
The sender of the transaction can be calculated as `0x3e266d3c3a70c238bdddafef1ba06fbd58958d70`. The address of the first contract deployed from the account is `rlp([sender, 0])` which equals `0xbEac00dDB15f3B6d645C48263dC93862413A222D`. This is how `BEACON_ROOTS_ADDRESS` is determined. Although this style of contract creation is not tied to any specific initcode like create2 is, the synthetic address is cryptographically bound to the input data of the transaction (e.g. the initcode).
```

### Block processing

At the start of processing any execution block where `block.timestamp >= FORK_TIMESTAMP` (i.e. before processing any transactions), call `BEACON_ROOTS_ADDRESS` as `SYSTEM_ADDRESS` with the 32-byte input of `header.parent_beacon_block_root`. This will trigger the `set()` routine of the beacon roots contract. This is a system operation and therefore:

> * the call must execute to completion, therefore the available gas can be considered as infinite

Reviewer: Why? 
I would rather specify some hard limit, like `30M` gas, as suggested by @yperbasis . I suspect most actual implementation will need to set a limit anyway.

For a genesis block with no existing parent beacon block root the 32 zero bytes are used as a root placeholder.

#### Beacon roots contract

> The beacon roots contract has two operations: `get` and `set`. The input itself is not used to determine which function to execute, for that the result of `caller` is used. If `caller` is equal to `SYSTEM_ADDRESS` then the operation to perform is `set`. Otherwise, `get`.

Reviewer 6: oh, funny. seems okay though

Reviewer 7: This somewhat seems to imply that this contract has an ABI (which it has not, at least not in current assembly). I would really like to do the ABI-like approach as in https://github.com/lightclient/4788asm/pull/5 since this will also be very helpful for solidity users (you can now `BeaconRootContract(address).get(timest)`.

Author: I think it is a bad idea to enshrine solidity ABI behavior. It is ubiquitous today, but in the future I expect other language will have different calling conventions. Solidity should directly support this contract like they do for `ecrecover`.

Reviewer 2: sounds like a hack imo

##### `get`

> * If `caller` is equal to `SYSTEM_ADDRESS`, the contract must revert.

Reviewer 6: This seems unnecessary. The outer of `get`/`set` does the routing *based* on the `caller` so by definition the `caller` won't be `SYSTEM_ADDRESS`

(same for parallel in `set`)

Author: good point - updated

* Callers provide the `timestamp` they are querying encoded as 32 bytes in big-endian format.
* If the input is not exactly 32 bytes, the contract must revert.
* Given `timestamp`, the contract computes the storage index in which the timestamp is stored by computing the modulo `timestamp % HISTORICAL_ROOTS_MODULUS` and reads the value.
* If the `timestamp` does not match, the contract must revert.
* Finally, the beacon root associated with the timestamp is accessed at returned to the user. It is stored at `timestamp % HISTORICAL_ROOTS_MODULUS + HISTORICAL_ROOTS_MODULUS`.

##### `set`

> * If `caller` is not equal to `SYSTEM_ADDRESS`, the contract must revert.

Reviewer 6: ditto

* Caller provides the parent beacon block root as calldata to the contract.

> * Set the storage value at `header.timestamp % HISTORICAL_ROOTS_MODULUS` to be `header.timestamp`

Reviewer 6: I kind of want to make the case for avoiding  a system-TX and just setting these storage values at the top of the block. Then the code is just `get` and we don't have to test/specify how to do system-TXs 

weaker held opinion than the deploy method

Author: I think if we're going to use a standard contract we should embrace it and allow it to have storing functionality.

Reviewer: Wrote this already in discord, posting here too for visibility. 

With the spec being 'system call', it means a reference client can just do a simple call. A client implementor can still choose not to, and instead bypass the whole call and update the values directly. 

If the eip standardizes 'direct update', however, and omits the system-update path from the contract, then we remove all optionality, and force more special-case code into clients

Therefore I think the 'system call' approach is the best way forward.

Reviewer: I think the EIP should mention that "direct update" is a valid/possible client implementation/optimization detail.

Reviewer 2: having a system-tx would be a good generic functionality to have in long term imo

    timestamp = storage.get(timestamp_idx)

    if timestamp != evm.calldata:
        evm.revert()

    root_idx = timestamp_idx + HISTORICAL_ROOTS_MODULUS
    root = storage.get(root_idx)
    
    evm.return(root)

def set():
    timestamp_idx = to_uint256_be(evm.timestamp) % HISTORICAL_ROOTS_MODULUS
    root_idx = timestamp_idx + HISTORICAL_ROOTS_MODULUS

    storage.set(timestamp_idx, evm.timestamp)
    storage.set(root_idx, evm.calldata)
```

##### Bytecode

The exact initcode to deploy is shared below.

```asm
push1 0x5a
dup1
push1 0x09
push0
codecopy
push0
return

caller
push20 0xfffffffffffffffffffffffffffffffffffffffe
eq
push1 0x42
jumpi

push1 0x20
calldatasize
eq
push1 0x24
jumpi

push0
push0
revert

jumpdest
push3 0x018000
push0
calldataload
mod
dup1
sload
push0
calldataload
eq
iszero
push1 0x3d
jumpi

push3 0x018000
add
sload
push0
mstore

jumpdest
push1 0x20
push0
return

jumpdest
timestamp
push3 0x018000
timestamp
mod
sstore
push0
calldataload
push3 0x018000
timestamp
mod
push3 0x018000
add
sstore
stop
```

#### Deployment

The beacon roots contract is deployed like any other smart contract. A special synthetic address is generated

> by working backwards from the desired deployment transaction:

Reviewer 6: This should be specified as "at the first block where  `block.timestamp >= FORK_TIMESTAMP`, this deploy TX is transparently inserted at the start of the block" right?

Right now the spec doesn't say if/when to deploy the bytecode via this method

Author: Submission of the deployment tx is permissionless and can occur at anytime before or after fork.

```json
{
  "type": "0x2",
  "chainId": "0x1",
  "nonce": "0x0",
  "to": null,
  "gas": "0xd4f8",
  "gasPrice": null,
  "maxPriorityFeePerGas": "0x9c7652400",

>   "maxFeePerGas": "0xe8d4a51000",

Reviewer 7: The semi-problem with this fake sig tx is (as raised by @holiman in a discussion we had), that if you want to deploy on other chains, all items are fixed, so also these gas fees, which might be problematic to deploy it on another chain.

Author: We could do make uint256 for `maxFeePerGas` if we use an EIP-1559 tx. Ultimately it isn't important how other chains deploy the opcode. The EIP specifies the mainnet behavior, but I assume the address will ultimately become a genesis configuration option much like the deposit contract. Chains could easily etch the code directly into the state by an irregular state transition or via genesis enshrinement.

Reviewer 7: Ok that actually makes a lot of sense!

Reviewer 7: (Small side note: we actually cannot do max uint256 since this violates the balance check of 1559. But this does not matter because you are right, other chains can just "expand" this EIP to just deploy it anywhere as a special state transition)

Author: good point 🫠

  "value": "0x0",

>   "input": "0x605a8060095f395ff33373fffffffffffffffffffffffffffffffffffffffe14604257602036146024575f5ffd5b620180005f350680545f351415603d576201800001545f525b60205ff35b42620180004206555f3562018000420662018000015500",

Reviewer 6: I might put many of these values behind a VAR -- e.g. input and the stubbed `r`, `s`

  "accessList": [],
  "v": "0x0",

>   "r": "0x539",

Reviewer 6: > Although this style of contract creation is not tied to any specific initcode like create2 is, the synthetic address is cryptographically bound to the input data of the transaction (e.g. the initcode).

and then we don't have to think about things like this.

>   "r": "0x539",

Author: >  rather than trying to use a TX of sorts but without gas or signature verification which requires a lot more exceptional logic

To be clear, this tx does still abide by gas and signature verification. The signature is constructed arbitrarily and therefore the sender is a one-time sender address (pk is not known).

--

In an effort to minimize protocol baggage, I think it is better to deploy the contract using standard methods. Once we are satisfied with the bytecode we'd like to deploy, it could be deployed by anyone. We simply have to agree on the address at which it is deployed. It doesn't have to be a synthetic tx, although I do slightly prefer this method as it is permissionless (anyone can fund the account and submit the tx).

  "s": "0x1337",
  "hash": "0x8ecfe5753922d27aa737597d946f638a15b7e3b5f74fef9ef8cf1b510a1af1cc"
  }
```

The sender of the transaction can be calculated as `0x01d0610058aC7AEF1887d8877ee7f04B7645Dc95`. The address of the first contract deployed from the account is `rlp([sender, 0])` which equals `0x502E02F5d91024A9AF0aB81fbF0a47Eb99a013aE`. Although this style of contract creation is not tied to any specific initcode like create2 is, the synthetic address is cryptographically bound to the input data of the transaction (e.g. the initcode).

### Block processing

At the start of processing any execution block where `block.timestamp >= FORK_TIMESTAMP` (i.e. before processing any transactions), call `BEACON_ROOTS_ADDRESS` as `SYSTEM_ADDRESS` with the 32-byte input of `header.parent_beacon_block_root`. This will trigger the `set()` routine of the beacon roots contract. This is a system operation and therefore:

* the call must execute to completion, therefore the available gas can be considered as infinite
* the call does not count against the block's gas limit
* the call does not follow the [EIP-1559](./eip-1559.md) burn semantics - no value should be transferred as part of the call

> * if no code exists at `BEACON_ROOTS_ADDRESS`, the call must fail silently

Reviewer 6: Assuming the prior section specifies when to deploy the bytecode, this bullet will become unnecessary

Author: As mentioned in other thread, I don't think it is important to specify the deployment. But I suppose the bullet is unnecessary because that is the semantics of an evm call anyways.

Reviewer: This bullet is unnecessary, other than as clarification: any non-value call to account with non-existing code has no discernible effect on state.  
If you prefix the bullet with `Note:` or `Clarification:`, it becomes apparent that this bullet does not intend to add any behavioural changes.

Reviewer 8: @holiman But shouldn't it then rather say:

* if no code exists at `BEACON_ROOTS_ADDRESS`, the call must succeed as with any other account without code

I find "fail" a bit misleading here.

Reviewer 9: I agree: fail or succeed depends on your point of view. Whole bullet is moot and should not exist
